diff --git a/arch/arm64/boot/dts/rockchip/rk3568-hm.dts b/arch/arm64/boot/dts/rockchip/rk3568-hm.dts
index ab8c8e0bfcf5..5c8652a52e89 100644
--- a/arch/arm64/boot/dts/rockchip/rk3568-hm.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3568-hm.dts
@@ -43,7 +43,7 @@
 		compatible = "gpio-leds";
 
 		led_user: led-0 {
-			gpios = <&gpio0 RK_PB7 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpio0 RK_PC0 GPIO_ACTIVE_LOW>;
 			function = LED_FUNCTION_HEARTBEAT;
 			color = <LED_COLOR_ID_BLUE>;
 			linux,default-trigger = "heartbeat";
